### PR TITLE
VMulCAddC-Replaced yaml files with header table

### PR DESCRIPTION
**Pull request by nitheshsrikanth-mcw**

### CODE REVIEW

File: tools/generate-vmulcaddc-test.py
Comment (dsharletg) [inline tools/generate-vmulcaddc-test.py:1]: Same comments as the other PRs

**Review by dsharletg — APPROVED** [review]

**Comment by nitheshsrikanth-mcw** [comment]
@dsharlet Thank you for your comment. I did attempt to update the kernel declarations as suggested, but I encountered an issue during the process that I wasn't able to resolve today. I will revisit the task tomorrow and work on resolving it then.

**Comment by nitheshsrikanth-mcw** [comment]
@dsharlet i have updated the kernel declarations. Thanks!

File: test/vmulcaddc-microkernel-tester.h [hunk 1]
```diff
@@ -255,3 +247,173 @@ class VMulCAddCMicrokernelTester {
   uint8_t qmax_{255};
   size_t iterations_{15};
 };
+
+#define XNN_TEST_VMULCADDC_ROW_DIV(ukernel, arch_flags, row_tile, channels_tile, datatype, params_type, init_params)   \
+  TEST(ukernel, ROW_div)                                                                                               \
+  {                                                                                                                    \
+    for (size_t rows = row_tile * 2; rows <= row_tile * 4; rows += row_tile) {                                         \
```
Comment (dsharletg) [inline test/vmulcaddc-microkernel-tester.h:254]: `TEST_REQUIRES_ARCH_FLAGS(arch_flags)`? Does this pass tests locally for you?

Comment (nitheshsrikanth-mcw) [reply test/vmulcaddc-microkernel-tester.h:255]: Thanks for pointing out the missing TEST_REQUIRES_ARCH_FLAGS(arch_flags). The issue with the illegal instruction for AVX-512 has now been resolved and now the tests are passing.

File: src/xnnpack/vmulcaddc.h [hunk 2]
```diff
@@ -26,52 +26,21 @@ extern "C" {
...
-DECLARE_F32_VMULCADDC_MINMAX_UKERNEL_FUNCTION(xnn_f32_vmulcaddc_minmax_ukernel_c2__scalar_2x)
-DECLARE_F32_VMULCADDC_MINMAX_UKERNEL_FUNCTION(xnn_f32_vmulcaddc_minmax_ukernel_c4__scalar_2x)
-
-#define DECLARE_F16_VMULCADDC_MINMAX_UKERNEL_FUNCTION(fn_name) \
+#define XNN_UKERNEL(arch_flags, fn_name, row_tile, channel_tile, datatype) \
   XNN_INTERNAL void fn_name(                                   \
       size_t m,                                                \
       size_t c,                                                \
-      const xnn_float16* x,                           \
+      const xnn_float16* x,                                    \
```
Comment (dsharlet) [inline src/xnnpack/vmulcaddc.h:36]: We should be able to use `datatype` here, and `#include` both headers with the one `XNN_UKERNEL` definition.

File: test/f16-vmulcaddc-minmax.cc [hunk 3]
```diff
@@ -4,629 +4,28 @@
...
-          .channel_tile(16)
-          .channels(channels)
-          .rows(rows)
-          .qmax(128)
-          .Test(xnn_f16_vmulcaddc_minmax_ukernel_c16__fma3_2x, xnn_init_f16_minmax_scalar_params);
-      }
-    }
-  }
-#endif  // XNN_ARCH_X86 || XNN_ARCH_X86_64
+#include "next_prime.h"
```
Comment (dsharlet) [inline test/f16-vmulcaddc-minmax.cc:17]: If this is really necessary, it should be in vmulcaddc-microkernel-tester.h. But I don't think it is.

File: test/f32-vmulcaddc-minmax.cc [hunk 4]
```diff
@@ -4,2877 +4,28 @@
...
-        .channel_tile(4)
-        .channels(channels)
-        .rows(rows)
-        .qmax(128)
-        .Test(xnn_f32_vmulcaddc_minmax_ukernel_c4__scalar_2x, xnn_init_f32_minmax_scalar_params);
-    }
-  }
-}
\ No newline at end of file
+#include "next_prime.h"
```
Comment (dsharlet) [inline test/f32-vmulcaddc-minmax.cc:17]: Same

**Comment by dsharlet** [comment]
Note merge conflicts

File: test/f32-vmulcaddc-minmax.cc (hunk 4 above)
Comment (nitheshsrikanth-mcw) [reply test/f32-vmulcaddc-minmax.cc]: Removed

File: test/f16-vmulcaddc-minmax.cc (hunk 3 above)
Comment (nitheshsrikanth-mcw) [reply test/f16-vmulcaddc-minmax.cc]: Didn't notice , Now it's been removed

File: scripts/generate-tests.sh [hunk 5]
```diff
@@ -251,9 +251,11 @@ tools/generate-raddexpminusmax-test.py --spec test/f32-raddexpminusmax.yaml --ou
 tools/generate-raddstoreexpminusmax-test.py --spec test/f16-raddstoreexpminusmax.yaml --output test/f16-raddstoreexpminusmax.cc &
 tools/generate-raddstoreexpminusmax-test.py --spec test/f32-raddstoreexpminusmax.yaml --output test/f32-raddstoreexpminusmax.cc &
 
-### Tests for VMulCAddC micro-kernels
-tools/generate-vmulcaddc-test.py --spec test/f16-vmulcaddc-minmax.yaml --output test/f16-vmulcaddc-minmax.cc &
-tools/generate-vmulcaddc-test.py --spec test/f32-vmulcaddc-minmax.yaml --output test/f32-vmulcaddc-minmax.cc &
+### Tests for VScaleExtExp micro-kernels
```
Comment (dsharlet) [inline scripts/generate-tests.sh:254]: What's going on here? Bad merge?

Comment (nitheshsrikanth-mcw) [reply scripts/generate-tests.sh]: Addressed

**Review by dsharlet — APPROVED** [review]

**Comment by dsharlet** [comment]
@nitheshsrikanth-mcw please take a look at the tests, for example:
```
/home/runner/work/XNNPACK/XNNPACK/src/f16-vmulcaddc/gen/f16-vmulcaddc-c8-minmax-neonfp16arith-2x.c:18:6: error: conflicting types for ‘xnn_f16_vmulcaddc_minmax_ukernel_c8__neonfp16arith_2x’; have ‘void(size_t,  size_t,  const xnn_float16 * restrict,  size_t,  const xnn_float16 * restrict,  xnn_float16 * restrict,  size_t,  const union xnn_f16_minmax_params * restrict)’ {aka ‘void(unsigned int,  unsigned int,  const xnn_float16 * restrict,  unsigned int,  const xnn_float16 * restrict,  xnn_float16 * restrict,  unsigned int,  const union xnn_f16_minmax_params * restrict)’}
   18 | void xnn_f16_vmulcaddc_minmax_ukernel_c8__neonfp16arith_2x(
```
But they are all failing.

**Comment by dsharlet** [comment]
Why would including next-prime.h in test/vmulcaddc-microkernel-tester.h address a compilation error in a microkernel?

Can you build and run this branch locally, in any configuration?

**Comment by nitheshsrikanth-mcw** [comment]
It's working for x86 now after updating the next prime header (the error was related to it). For ARM, we need to check, and I'll update this soon.

**Review by dsharlet — APPROVED** [review]

**Review by dsharlet — APPROVED** [review]

**Review by dsharlet — APPROVED** [review]

File: test/vmulcaddc-microkernel-tester.h [hunk 6]
```diff
@@ -255,3 +246,163 @@ class VMulCAddCMicrokernelTester {
   uint8_t qmax_{255};
   size_t iterations_{15};
 };
+
+#define XNN_TEST_VMULCADDC_ROW_DIV(ukernel, arch_flags, row_tile_, channel_tile_, datatype, params_type, init_params)  \
+  TEST(ukernel, ROW_div)                                                                                               \
+  {                                                                                                                    \
+    for (size_t rows_ = row_tile_ * 2; rows_ <= row_tile_ * 4; rows_ += row_tile_) {                                   \
```
Comment (dsharlet) [inline test/vmulcaddc-microkernel-tester.h:253]: This is missing TEST_REQUIRES_ARCH_FLAGS(arch_flags) in all tests.

**Review by dsharlet — APPROVED** [review]

**Pull request merged**